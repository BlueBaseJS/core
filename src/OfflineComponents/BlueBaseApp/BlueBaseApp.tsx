import { BlueBase, BlueBaseProgress, BootOptions } from '../../BlueBase';
import {
	BlueBaseAppError,
	BlueBaseAppErrorProps,
	BlueBaseAppLoading,
	BlueBaseAppLoadingProps,
} from '..';
import React, { useEffect, useState } from 'react';

import { BlueBaseProvider } from '../../Context';
import { IntlProvider } from '../../intl';
import { ThemeProvider } from '../../themes';
import { getComponent } from '../../getComponent';

const BlueBaseContent = getComponent('BlueBaseContent');

export interface BlueBaseAppProps extends Partial<BootOptions> {
	/** BlueBase Context */
	BB?: BlueBase;

	/** Component rendered when an error occurs */
	ErrorComponent: React.ComponentType<BlueBaseAppErrorProps>;

	/** Component rendered when an error occurs */
	LoadingComponent: React.ComponentType<BlueBaseAppLoadingProps>;

	/**
	 * If this prop is provided, this tree will be rendered instead of BlueBase's own view.
	 */
	children?: React.ReactNode;

	/**
	 * Used to locate this view in end-to-end tests.
	 */
	testID?: string;
}

/**
 * # ðŸš€ BlueBaseApp
 *
 * The main BlueBase app. This is the top level component in BlueBase. Takes care
 * of initialisation, and renders either children, or app with routing.
 *
 * ## Usage
 * ```jsx
 * <BlueBaseApp BB={BB} />
 * ```
 */
export const BlueBaseApp = (props: BlueBaseAppProps) => {
	const {
		ErrorComponent = BlueBaseAppError,
		LoadingComponent = BlueBaseAppLoading,
		children,
	} = props;

	const [bootCount, setBootCount] = useState(0);
	const [booting, setBooting] = useState(true);
	const [bootTrigger, setBootTrigger] = useState(true); // Setting to true to start boot
	const [progress, setProgress] = useState<BlueBaseProgress>({});

	const [BB] = useState(props.BB || new BlueBase());

	BB.reboot = async () => {
		setBootTrigger(true);
	};

	useEffect(() => {
		if (bootTrigger) {
			setBootTrigger(false);
		} else {
			return;
		}

		(async () => {
			await BB.boot({
				...props,
				onProgress: (p: BlueBaseProgress) => {
					setProgress(p);
					if (p.error) {
						setBooting(false);
					}
				},
			});
			setBootCount(bootCount + 1);
			setBooting(false);
		})();
	}, [bootTrigger]);

	if (booting) {
		return <LoadingComponent BB={BB} progress={progress} bootCount={bootCount} />;
	}

	if (progress.error) {
		return <ErrorComponent BB={BB} progress={progress} bootCount={bootCount} />;
	}

	return (
		<BlueBaseProvider key={`boot-${bootCount}`} value={BB}>
			<ThemeProvider>
				<IntlProvider>
					<BlueBaseContent>{children}</BlueBaseContent>
				</IntlProvider>
			</ThemeProvider>
		</BlueBaseProvider>
	);
};

const defaultProps: Partial<BlueBaseAppProps> = {
	ErrorComponent: BlueBaseAppError,
	LoadingComponent: BlueBaseAppLoading,
};

BlueBaseApp.defaultProps = defaultProps;
