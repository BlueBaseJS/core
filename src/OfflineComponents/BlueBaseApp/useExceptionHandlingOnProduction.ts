import { BlueBase } from '../../BlueBase';
import { isProduction } from '../../utils';

// Number of errors to save
const CACHE_SIZE = 5;

// If a successive error take longer than this time, we don't consider it as recursive
const RECURSIVE_PERIOD = 300;

let errors: Array<{
	error: any;
	time: number;
}> = [];

/**
 * If the app is in production mode, automatically reboots BB on exception.
 * Shows an error message if the app is stuck in infinite crash.
 * @param BB
 */
export function useExceptionHandlingOnProduction(BB: BlueBase) {
	let development = BB.Configs.getValue('development');

	if (development === undefined) {
		development = !isProduction();
	}

	/**
	 * Adds an error to the errors array, with the timestamp
	 * @param e
	 */
	function addError(e: any) {
		const arr = [
			...errors,
			{
				error: e,
				time: new Date().getTime(),
			},
		];

		errors = arr.length > CACHE_SIZE ? arr.slice(1) : arr;
	}

	/**
	 * Gets periods of time in successive errors
	 */
	function isInfinite() {
		if (errors.length < CACHE_SIZE) {
			return false;
		}

		for (let i = 1; i < errors.length; i++) {
			if (errors[i].time - errors[i - 1].time > RECURSIVE_PERIOD) {
				return false;
			}
		}

		return true;
	}

	function onError(e: any) {
		addError(e);

		if (!development && !isInfinite()) {
			BB.reboot();
		}
	}

	return { onError };
}
