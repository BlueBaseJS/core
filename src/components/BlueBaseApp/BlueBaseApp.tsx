import { BlueBase, BootOptions } from '../../BlueBase';
import React from 'react';
import { Text } from 'react-native';

export interface BlueBaseAppProps extends Partial<BootOptions> {
	BB?: BlueBase,
	children?: React.ReactNode,
}

interface BlueBaseAppState {

	booted: boolean;
	BB: BlueBase

	Component: React.ComponentType<any>
}

/**
 * ðŸš€ BlueBaseApp
 *
 * This is the Main App used to render at the top level.
 *
 * TODO: Add better docs
 */
// NOTE FOR DEVELOPERS:
// Don't use BlueBase context or any data saved in context (e.g. components)
// here. This is because the context may not be initialized yet.
export class BlueBaseApp extends React.Component<BlueBaseAppProps, BlueBaseAppState> {

	constructor(props: BlueBaseAppProps) {
		super(props);

		this.state = {
			BB: props.BB || new BlueBase(),
			Component: () => <Text>Loading</Text>,
			booted: false,
		};
	}

	async componentDidMount() {

		const BB = this.state.BB;

		try {
			const Component = await BB.boot(this.props);
			this.setState({
				Component,
				booted: BB.booted,
			});
		} catch (error) {
			// tslint:disable-next-line:no-console
			console.error(error);
		}
	}

	render() {

		const Component = this.state.Component;

		return <Component children={this.props.children} />;
	}
}