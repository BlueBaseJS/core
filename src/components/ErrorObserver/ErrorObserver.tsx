import { BlueBase } from '../../BlueBase';
import { BlueBaseContext } from '../../Context';
import { MaybeRenderPropChildren } from '../../utils';
import React from 'react';

const MISSING_ERROR = 'An unknown error occured.';

export interface ErrorObserverProps {
	error?: Error,
	checkError?: (props: ErrorObserverProps & any) => Error,
	errorComponent?: React.ComponentType<any>;
	children?: MaybeRenderPropChildren;
}

export interface ErrorObserverState {
	readonly error?: Error,
}

export class ErrorObserver extends React.PureComponent<ErrorObserverProps, ErrorObserverState> {

	static contextType = BlueBaseContext;

	public static defaultProps: Partial<ErrorObserverProps> = {
		checkError: (props) => props.error,
	};

	readonly state: ErrorObserverState = {
		error: (this.props.checkError) ? this.props.checkError(this.props) : undefined,
	};

	componentDidCatch(error: Error | null) {
		this.setState({
			error: error || new Error(MISSING_ERROR)
		});
	}

	render() {

		// FIXME: remove typecasting, added because current react typings don't seem to support this.context
		const BB: BlueBase = (this as any).context;

		const { error } = this.state;
		const { children } = this.props;

		if (error) {
			BB.Logger.error(error);
			const Error = this.props.errorComponent || BB.Components.ErrorState;
			return React.createElement(Error, { error });
		}

		// 'children' as a function, 'render prop' pattern
		if (typeof children === 'function') {
			return (children as any)();
		}

		return children;
	}
}
