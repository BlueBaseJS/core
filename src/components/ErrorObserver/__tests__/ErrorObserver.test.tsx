import { BlueBaseApp } from '../../BlueBaseApp';
import { ErrorObserver } from '../ErrorObserver';
import React from 'react';
import TestRenderer from 'react-test-renderer';
import { Text } from 'react-native';

beforeEach(() => {
	jest.resetModules();
});

const Bang = () => {
	throw Error('ğŸ’¥ Boom!');
};

const BangNull = () => {
	throw null;
};

describe('ErrorObserver', () => {

	test(`Snapshot ErrorObserver`, (done) => {
		const rendered = TestRenderer.create(
			<BlueBaseApp>
				<ErrorObserver>
					<Text>Hello</Text>
				</ErrorObserver>
			</BlueBaseApp>
		);

		setTimeout(() => {
			const tree = rendered.toJSON();
			expect((tree as any).children[0].children.join()).toBe('Hello');
			expect(tree).toMatchSnapshot();
			done();
		});
	});

	test(`Snapshot ErrorObserver after complete rendering`, (done) => {
		const rendered = TestRenderer.create(
			<BlueBaseApp>
				<ErrorObserver>
					<Text>Hello</Text>
				</ErrorObserver>
			</BlueBaseApp>
		);

		setTimeout(() => {
			const tree = rendered.toJSON();
			expect((tree as any).children[0].children.join()).toBe('Hello');
			expect(tree).toMatchSnapshot();
			done();
		});
	});

	test(`Snapshot ErrorObserver after complete rendering with error`, (done) => {
		const rendered = TestRenderer.create(
			<BlueBaseApp>
				<ErrorObserver
					error={{ name: '404', message: 'no page found' }}
					errorComponent={() => <Text>Error</Text>}
				>
					<Text>Hello</Text>
				</ErrorObserver>
			</BlueBaseApp>
		);

		setTimeout(() => {
			const tree = rendered.toJSON();
			expect((tree as any).children[0].children.join()).toBe('Error');
			expect(tree).toMatchSnapshot();
			done();
		});
	});

	test(`Snapshot ErrorObserver after complete rendering with child as function`, (done) => {
		const rendered = TestRenderer.create(
			<BlueBaseApp>
				<ErrorObserver>
				{
					() => <Text>Hello</Text>
				}
				</ErrorObserver>
			</BlueBaseApp>
		);

		setTimeout(() => {
			const tree = rendered.toJSON();
			expect((tree as any).children[0].children.join()).toBe('Hello');
			expect(tree).toMatchSnapshot();
			done();
		});
	});

	test(`should catch an error when thrown by a child component`, (done) => {
		const rendered = TestRenderer.create(
			<BlueBaseApp>
				<ErrorObserver>
					<Bang />
				</ErrorObserver>
			</BlueBaseApp>
		);

		setTimeout(() => {
			const tree = rendered.toJSON();
			expect((tree as any).children[0].children[0].children.join()).toBe('Error');
			expect((tree as any).children[0].children[1].children.join()).toBe('ğŸ’¥ Boom!');
			expect(tree).toMatchSnapshot();
			done();
		});
	});

	test(`should catch an error when thrown null by a child component`, (done) => {
		const rendered = TestRenderer.create(
			<BlueBaseApp>
				<ErrorObserver>
					<BangNull />
				</ErrorObserver>
			</BlueBaseApp>
		);

		setTimeout(() => {
			const tree = rendered.toJSON();
			expect((tree as any).children[0].children[0].children.join()).toBe('Error');
			expect((tree as any).children[0].children[1].children.join()).toBe('An unknown error occured.');
			expect(tree).toMatchSnapshot();
			done();
		});
	});

	test(`should not catch an error because check error is null`, (done) => {
		const rendered = TestRenderer.create(
			<BlueBaseApp>
				<ErrorObserver checkError={null as any} error={Error('Useless')} >
					<Text>Hello</Text>
				</ErrorObserver>
			</BlueBaseApp>
		);

		setTimeout(() => {
			const tree = rendered.toJSON();
			expect((tree as any).children[0].children.join()).toBe('Hello');
			expect(tree).toMatchSnapshot();
			done();
		});
	});


});
