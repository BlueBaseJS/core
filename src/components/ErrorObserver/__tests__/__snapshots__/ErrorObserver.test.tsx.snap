// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorObserver Snapshot ErrorObserver 1`] = `
<View>
  <Text>
    Hello
  </Text>
</View>
`;

exports[`ErrorObserver Snapshot ErrorObserver after complete rendering 1`] = `
<View>
  <Text>
    Hello
  </Text>
</View>
`;

exports[`ErrorObserver Snapshot ErrorObserver after complete rendering with child as function 1`] = `
<View>
  <Text>
    Hello
  </Text>
</View>
`;

exports[`ErrorObserver Snapshot ErrorObserver after complete rendering with error 1`] = `
<View>
  <Text>
    Error
  </Text>
</View>
`;

exports[`ErrorObserver should catch an error when thrown by a child component 1`] = `
<View>
  <View
    style={
      Object {
        "alignItems": "center",
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "rgba(0, 0, 0, 0.87)",
            "fontFamily": "Roboto",
            "fontSize": 20,
            "fontWeight": "500",
          },
          Object {
            "marginBottom": 8,
            "textAlign": "center",
          },
        ]
      }
    >
      Error
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "rgba(0, 0, 0, 0.87)",
            "fontFamily": "Roboto",
            "fontSize": 16,
            "fontWeight": "400",
          },
          Object {
            "textAlign": "center",
          },
        ]
      }
    >
      ðŸ’¥ Boom!
    </Text>
  </View>
</View>
`;

exports[`ErrorObserver should catch an error when thrown null by a child component 1`] = `
<View>
  <View
    style={
      Object {
        "alignItems": "center",
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "rgba(0, 0, 0, 0.87)",
            "fontFamily": "Roboto",
            "fontSize": 20,
            "fontWeight": "500",
          },
          Object {
            "marginBottom": 8,
            "textAlign": "center",
          },
        ]
      }
    >
      Error
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "rgba(0, 0, 0, 0.87)",
            "fontFamily": "Roboto",
            "fontSize": 16,
            "fontWeight": "400",
          },
          Object {
            "textAlign": "center",
          },
        ]
      }
    >
      An unknown error occured.
    </Text>
  </View>
</View>
`;

exports[`ErrorObserver should not catch an error because check error is null 1`] = `
<View>
  <Text>
    Hello
  </Text>
</View>
`;
